### ISSUE TITLE
Add a Git styleguide

### ISSUE THREAD

**Issue by duggiefresh** [DockYard/styleguides#41]
Some references:

https://web.archive.org/web/20150204104601/http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html

http://chris.beams.io/posts/git-commit/

**Comment by duggiefresh** [comment]
Closing for now

**Comment by michaeldupuisjr** [comment]
I might be in favor of re-opening this as "version control" styleguide.

On a number of projects, where many different tools can be used (include JIRA), I've definitely seen teams develop conventions in their commit messages. For example, in an ongoing project, we reference the ticket number in the first line of the commit like so:

`[JIRA-219] Title of the commit...`

These conventions come in pretty handy when Github isn't the source of truth and you need to find commits that can't reference the Github issue # in the message.

I also think it's important for us to continually be improving our documentation and commit messages, as this facilitates the handoff to clients, and also makes us better programmers (since we need to articulate what our commit is, it pushes us to create more targeted, modular commits.

**Comment by duggiefresh** [comment]
Good call, Mike. 👍  Especially, since we've been working alongside other issue trackers.

Furthermore, Rowan just asked about what we write in commit messages, so this will serve as a decent guideline.

**Comment by michaeldupuisjr** [comment]
Items of interest:

- [ ] `rebase` vs `merge`
- [ ] When to `squash` commits vs. having multiple commits in a single PR

**Comment by sukima** [comment]
I am a huge fan of Git and good Git commit messages with clear Git histories. I was about to try tackling some of this but felt the list above was a bit ambiguous. Could these items be expanded on just enough for a contributor to develop an outline for each line item?

+ *We use git* – Is this just a statement or an overview of Git best practices expanding in further sections below?
+ *rebase vs merge* – What is the standard here?
  + Is there a preferred strategy (for example *prefer rebase over merge*)?
  + Or was this just a section to describe the methods used to accomplish either a rebase or merge?
  + Would this section describe [workflow](https://www.atlassian.com/git/articles/git-team-workflows-merge-or-rebase/)?
+ *When to squash commits vs. having multiple commits in a single PR* – This might be more an opinion based on the teams' preferences. Is there any general advise on this topic that someone outside DockYard could contribute to?
+ *Commit message conventions when working with non-Github tracking tools* – It would help to have a list of conventions. Hard for non-team members to construct this section of the guides.
